### PR TITLE
#666 fix Warning: Class __PHP_Incomplete_Class has no unserializer in…

### PR DESCRIPTION
**Pull request by SirSnyder**
… Unknown on line 0

### CODE REVIEW

File: src/prepare.c [hunk 1]
```diff
@@ -576,13 +588,20 @@ static inline void pthreads_prepare_closures(pthreads_object_t *thread) {
...
+			pthreads_prepared_entry_internal(thread, entry, 0);
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	ZEND_HASH_FOREACH_STR_KEY_PTR(PTHREADS_CG(thread->creator.ls, class_table), name, entry) {
+		if ((prepared = zend_hash_find_ptr(PTHREADS_CG(thread->local.ls, class_table), name)) && prepared->type != ZEND_INTERNAL_CLASS) {
+			pthreads_prepared_entry_static_members(thread, entry, prepared);
 		}
-	} ZEND_HASH_FOREACH_END(); 
+	} ZEND_HASH_FOREACH_END();
```
Comment (tpunt) [inline src/prepare.c]: This is a bit of a nitpick, but if you inverse the loops, then you can skip a check on each iteration:
```c
ZEND_HASH_FOREACH_STR_KEY_PTR(PTHREADS_CG(thread->local.ls, class_table), name, entry) {
    if (entry->type != ZEND_INTERNAL_CLASS) {
        pthreads_prepared_entry_static_members(
            thread,
            zend_hash_find_ptr(PTHREADS_CG(thread->creator.ls, class_table), name),
            entry
        );
    }
} ZEND_HASH_FOREACH_END();
```

I cannot immediately see anything wrong with this, since a child thread will only ever have classes that exist in the parent thread.

**Comment by tpunt** [comment]
This doesn't cover all cases, it seems. For example, the following errors:
```php
class SecondClass extends FirstClass {}
class FirstClass extends \Threaded
{
	public static $prop = [];
}
FirstClass::$prop[] = new SecondClass();
$thread = new class extends Thread {
    public function run() {
        var_dump(FirstClass::$prop);
    }
};
$thread->start() && $thread->join();
```

This can easily be changed by not copying the statics when preparing parent classes in `pthreads_copy_entry`:
```c
prepared->parent = pthreads_prepared_entry_internal(thread, candidate->parent, 0);
```

There could be other edge cases around the preparing of classes, but I haven't had the time to check for them.

**Comment by SirSnyder** [comment]
@tpunt both fixed and I've added your example as test.

**Comment by tpunt** [comment]
Merged, thanks!
